server/blob/blob-memory/src/main/java/org/apache/james/blob/memory/MemoryBlobStore.java

```diff
@@ -28,50 +28,79 @@
 
-    @Override
-    public CompletableFuture<BlobId> save(byte[] data) {
-        Preconditions.checkNotNull(data);
-        BlobId blobId = factory.forPayload(data);
+        private final ConcurrentHashMap<BlobId, byte[]> blobs;
+        private final BlobId.Factory factory;
 
-        blobs.put(blobId, data);
+        MemoryBlobStoreImpl(BlobId.Factory factory) {
+            this.factory = factory;
+            blobs = new ConcurrentHashMap<>();
+        }
 
-        return CompletableFuture.completedFuture(blobId);
-    }
+        @Override
+        public CompletableFuture<BlobId> save(byte[] data) {
+            Preconditions.checkNotNull(data);
+            BlobId blobId = factory.forPayload(data);
 
-    @Override
-    public CompletableFuture<BlobId> save(InputStream data) {
-        Preconditions.checkNotNull(data);
-        try {
-            byte[] bytes = IOUtils.toByteArray(data);
-            return save(bytes);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+            blobs.put(blobId, data);
+
+            return CompletableFuture.completedFuture(blobId);
+        }
+
+        @Override
+        public CompletableFuture<BlobId> save(InputStream data) {
+            Preconditions.checkNotNull(data);
+            try {
+                byte[] bytes = IOUtils.toByteArray(data);
+                return save(bytes);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        @Override
+        public CompletableFuture<byte[]> readBytes(BlobId blobId) {
+            return CompletableFuture.completedFuture(retrieveStoredValue(blobId));
+        }
+
+        @Override
+        public InputStream read(BlobId blobId) {
+            return new ByteArrayInputStream(retrieveStoredValue(blobId));
+        }
+
+        private byte[] retrieveStoredValue(BlobId blobId) {
+            return blobs.getOrDefault(blobId, new byte[]{});
         }
     }
 
+    private final MetricFactory metricFactory;
+    private final MemoryBlobStoreImpl memoryBlobStoreImpl;
+
+    public MemoryBlobStore(MetricFactory metricFactory, BlobId.Factory factory) {
+        this.metricFactory = metricFactory;
+        this.memoryBlobStoreImpl = new MemoryBlobStoreImpl(factory);
+    }
+
     @Override
-    public CompletableFuture<byte[]> readBytes(BlobId blobId) {
-        return CompletableFuture.completedFuture(retrieveStoredValue(blobId));
+    public BlobStore blobStoreImpl() {
+        return memoryBlobStoreImpl;
     }
 
     @Override
-    public InputStream read(BlobId blobId) {
-        return new ByteArrayInputStream(retrieveStoredValue(blobId));
+    public MetricFactory metricFactory() {
+        return metricFactory;
     }
 
-    private byte[] retrieveStoredValue(BlobId blobId) {
-        return blobs.getOrDefault(blobId, new byte[]{});
+    @Override
+    public String metricNamePrefix() {
+        return "MemoryBlobStore:";
```

Comment:
The 'metric prefix' should be the same for all blob stores IMO.

Reply:
we don't want to distinguish BlobStore impl in grafana board?